src/Functions/MatchImpl.h

```diff
@@ -391,22 +394,248 @@ struct MatchImpl
+        PaddedPODArray<UInt8> & res)
     {
-        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Function '{}' doesn't support non-constant needle argument", name);
+        const size_t haystack_size = haystack_offsets.size();
+
+        if (haystack_size != needle_offset.size())
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Function '{}' unexpectedly received a different number of haystacks and needles", name);
+
+        if (start_pos_ != nullptr)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Function '{}' doesn't support start_pos argument", name);
+
+        if (haystack_offsets.empty())
+            return;
+
+        String required_substr;
+        bool is_trivial;
+        bool required_substring_is_prefix; /// for `anchored` execution of the regexp.
+
+        size_t prev_haystack_offset = 0;
+        size_t prev_needle_offset = 0;
+
+        for (size_t i = 0; i < haystack_size; ++i)
+        {
+            const auto * const cur_haystack_data = &haystack_data[prev_haystack_offset];
+            const size_t cur_haystack_length = haystack_offsets[i] - prev_haystack_offset - 1;
+
+            const auto * const cur_needle_data = &needle_data[prev_needle_offset];
+            const size_t cur_needle_length = needle_offset[i] - prev_needle_offset - 1;
+
+            const auto & needle = String(
+                    reinterpret_cast<const char *>(cur_needle_data),
+                    cur_needle_length);
+
+            if (like && impl::likePatternIsSubstring(needle, required_substr))
+            {
+                if (required_substr.size() > cur_haystack_length)
+                    res[i] = negate;
+                else
+                {
+                    Searcher searcher(required_substr.data(), required_substr.size(), cur_haystack_length);
+                    const auto * match = searcher.search(cur_haystack_data, cur_haystack_length);
+                    res[i] = negate
+                        ^ (match != cur_haystack_data + cur_haystack_length);
+                }
+            }
+            else
+            {
+                // each row is expected to contain a different like/re2 pattern
+                // --> bypass the regexp cache, instead construct the pattern on-the-fly
+                const int flags = Regexps::buildRe2Flags<true, case_insensitive>();
+                const auto & regexp = Regexps::Regexp(Regexps::createRegexp<like>(needle, flags));
+
+                regexp.getAnalyzeResult(required_substr, is_trivial, required_substring_is_prefix);
+
+                if (required_substr.empty())
+                {
+                    if (!regexp.getRE2()) /// An empty regexp. Always matches.
+                    {
+                        res[i] = 1;
+                    }
+                    else
+                    {
+                        res[i] = negate
+                            ^ regexp.getRE2()->Match(
```

Comment:
I would prefer syntax like
```
auto match = regexp.getRE2()->Match(
 ...
res[i] = negate ^ match;
```

But would not insist :) The current code is well readable

Reply:
The style was actually copypasted from the existing code paths but, yeah, I can change it to your proposal. Surely provides a better experience when debugging.